README
Usage: gpw n, to generate n passwords

Notes for those compiling from source
++++++++++++++++++++++++++++++++++++++++++++
If compiling from source, you need a C++ compiler. I had to install the package

gcc-c++

in my case even though I already had gcc installed.

Also note that a "bare word" dictionary is assumed to be at

 /usr/share/dict/linux.words

Update the Makefile if yours is in a different path.


HISTORY
11/28/1998 DrJ password generator fetched from http://www.lille.com/tvvtools.html
3/15/13 DrJ fixed up a couple things and got it working on CentOS 6
10/15/2017 The Lightning Stalker modified to generate stronger passwords


loadtris.c
/* This program counts trigraphs in all the dictionaries specified
   and outputs a C++ include file of constants for gpw.C to use.

   Change to a C include file by changing "const" to "static."
   Output of this program needs postprocessing to eliminate comma closebrace,
   see the makefile for gpw.

   THVV 6/94 Coded
   */

#include "stdio.h"
#include "stdlib.h"

int tris[26][26][26];			/* Trigraph frequencies */
int duos[26][26];				/* Bigraph frequencies */
int sing[26];					/* Letter frequencies */
long max = 0;					/* largest triraph count */
int m1, m2, m3;					/* coords of largest count */
long sigma = 0;					/* Total letters */

FILE *fp;

int main (int argc, char ** argv) {
	char buf[100];
	int j;
	int k1, k2, k3;
	int c1, c2, c3;
	char s1[2], s2[2], s3[2];
	int argno, nfiles;

	for (c1=0; c1 < 26; c1++) {	/* Initialize arrays to zero */
		sing[c1] = 0;
		for (c2=0; c2 < 26; c2++) {
			duos[c1][c2] = 0;
			for (c3=0; c3 < 26; c3++) {
				tris[c1][c2][c3] = 0;
			}
		}
	}
	s1[1] = '\0';
	s2[1] = '\0';
	s3[1] = '\0';
	nfiles = 0;					/* count of files read */

	if (argc < 2) {
		printf (" USAGE: loadtris /usr/dict/words ...");
		exit (1);
	}
	for (argno = 1; argno < argc; argno++) {
		if ((fp = fopen (argv[argno], "r")) == NULL) {
			printf ("** file %s not found\n", argv[argno]);
			break;
		}
		nfiles++;
		while (fgets (buf, sizeof (buf), fp)) {
			j = 0;					/* j indexes the input */
			k2 = -1;				/* k1, k2 are coords of previous letter */
			k1 = -1;
			while (buf[j]) {		/* until we find the null char.. */
				k3 = buf[j];		/* Pick out a letter from the input */
				if (k3 > 'Z') {
					k3 = k3 - 'a';	/* map from a-z to 0-25 */
				}
				else {
					k3 = k3 - 'A';	/* map from A-Z to 0-25 */
				}
				if (k3 >= 0 && k3 <= 25) { /* valid subscript? */
					if (k1 >= 0) { /* do we have 3 letters? */
						tris[k1][k2][k3]++;	/* count */
						sigma++;			/* grand total */
						if (tris[k1][k2][k3] > max) {
							max = tris[k1][k2][k3];
							m1 = k1; /* note largest cell.. */
							m2 = k2; /* .. for interest */
							m3 = k3;
						}
					}
					if (k2 >= 0) {
						duos[k2][k3]++;	/* count 2-letter pairs */
					}
					sing[k3]++;		/* count single letter frequency */
					k1 = k2;		/* shift over */
					k2 = k3;
				}
				j++;
			}						/* while buf[j] */
		}							/* while fgets */
		fclose (fp);
	}							    /* for argno */

	if (nfiles) {				    /* find any input? */
		printf ("/* BEGIN INCLUDE FILE .. trigram.h */\n"); /* Multics style */
		printf ("\n");
		printf ("static long sigma = %ld;\n", sigma);
		/* (for my /usr/dict/words it is 125729, fits in a long) */

		/* For interest print out the most frequent entry. */
		/* (for my /usr/dict/words it is 863 = ATE, showing that a short works OK) */
		s1[0] = m1 + 'a';
		s2[0] = m2 + 'a';
		s3[0] = m3 + 'a';
		printf ("static short maxcell = %ld; /* %s%s%s */\n", max, s1, s2, s3);

		printf ("static long sing[26] = {");
		for (c1=0; c1 < 26; c1++) {
			printf ("%d, ", sing[c1]);
		}
		printf ("};\n");				/* oops, ends in comma closebrace */

		printf ("static short duos[26][26] = {");
		for (c1=0; c1 < 26; c1++) {
			s1[0] = c1+'A';
			printf ("\n/* %s */ ", s1);
			for (c2=0; c2 < 26; c2++) {
				printf ("%d, ", duos[c1][c2]);
			}
		}
		printf ("};\n");				/* oops, ends in comma closebrace */

		printf ("static short tris[26][26][26] = {");
		for (c1=0; c1 < 26; c1++) {
			for (c2=0; c2 < 26; c2++) {
				s1[0] = c1+'A';
				s2[0] = c2+'A';
				printf ("\n/* %s %s */ ", s1, s2);
				for (c3=0; c3 < 26; c3++) {
					printf ("%d, ", tris[c1][c2][c3]);
				}
			}
		}
		printf ("};\n");				/* comma closebrace again, fix later */
		printf ("/* END   INCLUDE FILE .. trigram.h */\n");
	}
	exit (0);
}



gpw.c
/* GPW - Generate pronounceable passwords
   This program uses statistics on the frequency of three-letter sequences
   in your dictionary to generate passwords.  The statistics are in trigram.h,
   generated there by the program loadtris.  Use different dictionaries
   and you'll get different statistics.

   This program can generate every word in the dictionary, and a lot of
   non-words.  It won't generate a bunch of other non-words, call them the
   unpronounceable ones, containing letter combinations found nowhere in
   the dictionary.  My rough estimate is that if there are 10^6 words,
   then there are about 10^9 pronounceables, out of a total population
   of 10^11 8-character strings.  I base this on running the program a lot
   and looking for real words in its output.. they are very rare, on the
   order of one in a thousand.

   This program uses "drand48()" to get random numbers, and "srand48()"
   to set the seed to the microsecond part of the system clock.  Works
   for AIX C++ compiler and runtime.  Might have to change this to port
   to another environment.

   The best way to use this program is to generate multiple words.  Then
   pick one you like and transform it with punctuation, capitalization,
   and other changes to come up with a new password.

   THVV 6/1/94 Coded

   Original author unknown - DrJ 3/15/13
   Copyright 2013, DrJ,  Artistic License 2.0, http://opensource.org/licenses/Artistic-2.0
   */

#include "trigram.h"
#include "stdio.h"
#include "stdlib.h"
#include "math.h"
#include <sys/types.h>
/* following for SysV. */
/* #include <bsd/sys/time.h> */
/* following for BSD */
#include <sys/time.h>

int rand_lim(int limit) {
/* return a random number between 0 and limit inclusive. */
  int divisor = RAND_MAX/(limit);
  int retval;

  do {
    retval = rand() / divisor;
  } while (retval > limit);

  return retval;
}

int main (int argc, char ** argv) {
	int password_length;		/* how long should each password be */
	int n_passwords;			/* number of passwords to generate */
	int pwnum;					/* number generated so far */
	int c1, c2, c3;				/* array indices */
	long sumfreq;				/* total frequencies[c1][c2][*] */
	double pik;					/* raw random number in [0.1] from drand48() */
	long ranno;					/* random number in [0,sumfreq] */
	long sum;					/* running total of frequencies */
	char password[100];			/* buffer to develop a password */
	int nchar;					/* number of chars in password so far */
	struct timeval systime;		/* time reading for random seed */
	struct timezone tz;			/* unused arg to gettimeofday */

	password_length = 8;		/* Default value for password length. Originally was 8 */
	n_passwords = 10;			/* Default value for number of pws to generate */

    gettimeofday (&systime, &tz); /* Read clock. */
	srand48 (systime.tv_usec);    /* Set random seed. */
	srand (uint(systime.tv_usec)); /* Seed for rand() */

	if (argc > 1) {				/* If args are given, convert to numbers. */
		n_passwords = atoi (&argv[1][0]);
		if (argc > 2) {
			password_length = atoi (&argv[2][0]);
		}
	}
	if (argc > 3 || password_length > 99 ||
		password_length < 0 || n_passwords < 0) {
		printf (" USAGE: gpw [npasswds] [pwlength]\n");
		exit (4);
	}

	/* Pick a random starting point. */
	/* (This cheats a little; the statistics for three-letter
       combinations beginning a word are different from the stats
       for the general population.  For example, this code happily
       generates "mmitify" even though no word in my dictionary
       begins with mmi. So what.) */
	for (pwnum=0; pwnum < n_passwords; pwnum++) {
		pik = drand48 ();		/* random number [0,1] */
		sumfreq = sigma;		/* sigma calculated by loadtris */
		ranno = pik * sumfreq;	/* Weight by sum of frequencies. */
		sum = 0;
		for (c1=0; c1 < 26; c1++) {
			for (c2=0; c2 < 26; c2++) {
				for (c3=0; c3 < 26; c3++) {
					sum += tris[c1][c2][c3];
					if (sum > ranno) { /* Pick first value */
						password[0] = 'a' + c1;
						password[1] = 'a' + c2;
						password[2] = 'a' + c3;
						c1 = c2 = c3 = 26; /* Break all loops. */
					}			/* if sum */
				}				/* for c3 */
			}					/* for c2 */
		}						/* for c1 */

		/* Do a random walk. */
		nchar = 3;				/* We have three chars so far. */
		while (nchar < password_length) {
			password[nchar] = '\0';
			password[nchar+1] = '\0';
			c1 = password[nchar-2] - 'a'; /* Take the last 2 chars */
			c2 = password[nchar-1] - 'a'; /* .. and find the next one. */
			sumfreq = 0;
			for (c3=0; c3 < 26; c3++)
				sumfreq += tris[c1][c2][c3];
			/* Note that sum < duos[c1][c2] because
			   duos counts all digraphs, not just those
			   in a trigraph. We want sum. */
			if (sumfreq == 0) { /* If there is no possible extension.. */
				break;	/* Break while nchar loop & print what we have. */
			}
			/* Choose a continuation. */
			pik = drand48 ();
			ranno = pik * sumfreq;	/* Weight by sum of frequencies for row. */
			sum = 0;
			for (c3=0; c3 < 26; c3++) {
				sum += tris[c1][c2][c3];
				if (sum > ranno) {
					password[nchar++] = 'a' + c3;
					c3 = 26;	/* Break the for c3 loop. */
				}
			}					/* for c3 */
		}							/* while nchar */
		/* Insert random number or symbol */
		nchar = rand_lim(password_length);
		password[nchar] = '!' + rand_lim(31);
		/* Change case of random character */
		int nchar_prev = nchar;
		do {
			nchar = rand_lim(password_length);
		} while (nchar == nchar_prev);
		password[nchar] = password[nchar] - 32;
		printf ("%s\n", password);
	}								/* for pwnum */
	exit (0);
}
